### PR TITLE
Fancy Calling Conventions: Default/Optional Arguments Milestone

### PR DESCRIPTION
**Pull request by aidandenlinger**
Implementation of default/optional arguments. Notable refactor of code - in `type-check.ts`, `GlobalTypeEnv` is extended (see `design-callingconventions.md` for more details).

cc teammates @enagola @Prajwal-YR

### CODE REVIEW

File: ir.ts [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
+import {Type, BinOp, UniOp } from './ast';
 
 export type Program<A> = { a?: A, funs: Array<FunDef<A>>, inits: Array<VarInit<A>>, classes: Array<Class<A>>, body: Array<BasicBlock<A>> }
 
 export type Class<A> = { a?: A, name: string, fields: Array<VarInit<A>>, methods: Array<FunDef<A>>}
 
 export type VarInit<A> = { a?: A, name: string, type: Type, value: Value<A> }
 
+export type Parameter<A> = { name: string, type: Type }
+
```
Comment (GV1028) [inline ir.ts]: Clean up new line changes

File: package.json [hunk 2]
```diff
@@ -7,6 +7,7 @@
     "build-web": "webpack && cp \"index.html\" \"build/\"",
     "build-parse": "tsc parsestart.ts --esModuleInterop --moduleResolution node --outDir cli/",
     "test": "set TS_NODE_COMPILER_OPTIONS={\"module\": \"commonjs\" } && mocha -r ts-node/register 'tests/**/*.test.ts'",
+    "test-callconv": "env TS_NODE_COMPILER_OPTIONS='{\"module\": \"commonjs\" }' node node_modules/.bin/mocha -r ts-node/register --reporter mochawesome 'tests/callingconventions.test.ts'",
```
Comment (GV1028) [inline package.json]: Nice! Might want to run this by the testing team as well.

**Review by GV1028 — COMMENTED** [review]
Overall looks good! You might want to talk to the closures group as well regarding any interactions, but I don't see any problems now. Also, try adding a few more test cases where the parameters are computed values.

File: ir.ts (hunk 1 above)
Comment (aidandenlinger) [reply ir.ts]: Cleaned up.

File: package.json (hunk 2 above)
Comment (aidandenlinger) [reply package.json]: We've removed this shortcut for now to not pollute package.json with only our tests, will run this by testing team to see if they think shortcuts like these should be adopted universally.

**Comment by aidandenlinger** [comment]
We discussed with the closure team in class on Tuesday and agree that with their design, we don't have any major problems. We've also added more test cases.

**Comment by aidandenlinger** [comment]
Note: when merging our branch with upstream, we've found a conflict with the destructuring group and how they're inserting a `Range$hasNext()` function into the AST during lowering, which results in a method call (this call of `hasNext` on a Range object) entering our code for a function call, which does not look at class information. We're in contact with the destructuring team to resolve this issue, but as of right now this PR isn't ready to be merged into upstream.

Our current work on this is at <https://github.com/aidandenlinger/chocopy-wasm-compiler-B/tree/week7debug>, but is incomplete. 

(We also had a conflict with the built-in libraries group and how they removed `builtin1` and `builtin2`, but we've resolved this).

**Comment by Prajwal-YR** [comment]
The PR has been updated to include changes from upstream. However, two range testcases designed by destructuring assignments team are failing. We have asked them to look into it. @techno96 @yalhessi @GV1028 @jpolitz 

CC: @aidandenlinger @enagola

**Comment by aidandenlinger** [comment]
Closing in favor of #75, which has this functionality as well as named arguments and does not have upstream conflicts. It would be easier on our team to only keep our latest code up to date with upstream, rather than maintaining two branches. If it would be preferable to only merge default arguments, however, just leave a comment and let us know.

cc @enagola @Prajwal-YR

**Pull request closed without merging**